pkg/jtracer/jtracer.go

```diff
@@ -14,16 +14,103 @@
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/go-logr/zapr"
+	"github.com/opentracing/opentracing-go"
+	"go.opentelemetry.io/otel"
+	otbridge "go.opentelemetry.io/otel/bridge/opentracing"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
 
 type JTracer struct {
-	OT opentracing.Tracer
+	OT   opentracing.Tracer
+	OTEL *sdktrace.TracerProvider
+	log  *zap.Logger
 }
 
-func OT(t opentracing.Tracer) JTracer {
-	return JTracer{OT: t}
+var once sync.Once
+
+func New() JTracer {
+	jt := JTracer{}
+	zc := zap.NewDevelopmentConfig()
+	zc.Level = zap.NewAtomicLevelAt(zapcore.Level(-8)) // level used by OTEL's Debug()
+	logger, err := zc.Build()
+	if err != nil {
+		panic(err)
+	}
+	otel.SetLogger(zapr.NewLogger(logger))
+
+	opentracingTracer, otelTracerProvider := jt.initBoth()
+
+	return JTracer{OT: opentracingTracer, OTEL: otelTracerProvider, log: logger}
 }
 
 func NoOp() JTracer {
-	return JTracer{OT: opentracing.NoopTracer{}}
+	return JTracer{OT: opentracing.NoopTracer{}, OTEL: &sdktrace.TracerProvider{}}
+}
+
+// initBoth initializes OpenTelemetry SDK and uses OTel-OpenTracing Bridge
+func (jt JTracer) initBoth() (opentracing.Tracer, *sdktrace.TracerProvider) {
+	traceExporter, err := newExporter()
+	if err != nil {
+		jt.log.Sugar().Fatalf("failed to create exporter", zap.Any("error", err))
+	}
+
+	// Register the trace exporter with a TracerProvider, using a batch
+	// span processor to aggregate spans before export.
+	bsp := sdktrace.NewBatchSpanProcessor(traceExporter, sdktrace.WithBlocking())
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSpanProcessor(bsp),
+		sdktrace.WithResource(resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("otlp"),
+		)),
+	)
+
+	once.Do(func() {
+		otel.SetTextMapPropagator(
+			propagation.NewCompositeTextMapPropagator(
+				propagation.TraceContext{},
+				propagation.Baggage{},
+			))
+	})
+
+	// Use the bridgeTracer as your OpenTracing tracer(otTrace).
+	otTracer, _ := otbridge.NewTracerPair(tracerProvider.Tracer(""))
+
+	return otTracer, tracerProvider
+}
+
+// newExporter returns a console exporter.
+func newExporter() (sdktrace.SpanExporter, error) {
+	client := otlptracegrpc.NewClient(
+		otlptracegrpc.WithInsecure(),
+	)
+	// Set up a otlp-trace exporter
+	traceExporter, err := otlptrace.New(context.Background(), client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create trace exporter: %w", err)
+	}
+
+	return traceExporter, nil
+}
+
+// Shutdown the tracerProvider to clean up resources
+func (jt JTracer) Close(ctx context.Context) {
+	err := jt.OTEL.Shutdown(ctx)
+	if err != nil {
+		log.Fatalf("Error shutting down tracer provider: %v", err)
+	}
```

Comment:
```suggestion
func (jt JTracer) Close(ctx context.Context) error {
	return jt.OTEL.Shutdown(ctx)
```

Reply:
I wondered how I can close this if we use `trace.TracerProvider` interface.

Reply:
There are several ways:
  1. you can return Closer, similar to how Jaeger SDK's creation returns `(opentracing.Tracer, io.Closer, error)`
  2. you can store the `provider.Shutdown` function as a private member of JTracer and invoke it from this Close method

Reply:
Thanks @yurishkuro for your masterpiece. 😄